quickwit/quickwit-rest-client/src/lib.rs

```diff
@@ -17,6 +17,124 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
+use std::path::Path;
+use std::{io, mem};
+
+use bytes::Bytes;
+use tokio::fs::File;
+use tokio::io::{AsyncBufReadExt, AsyncRead, BufReader};
+
 pub mod error;
 pub mod models;
 pub mod rest_client;
+
+pub(crate) struct BatchLineReader {
+    buf_reader: BufReader<Box<dyn AsyncRead + Send + Sync + Unpin>>,
+    buffer: Vec<u8>,
+    alloc_num_bytes: usize,
+    batch_num_bytes: usize,
+}
+
+impl BatchLineReader {
+    pub async fn from_file(filepath: &Path, batch_num_bytes: usize) -> io::Result<Self> {
+        let file = File::open(&filepath).await?;
+        Ok(Self::new(Box::new(file), batch_num_bytes))
+    }
+
+    pub fn from_stdin(batch_num_bytes: usize) -> Self {
+        Self::new(Box::new(tokio::io::stdin()), batch_num_bytes)
+    }
+
+    pub fn new(reader: Box<dyn AsyncRead + Send + Sync + Unpin>, batch_num_bytes: usize) -> Self {
+        let alloc_num_bytes = batch_num_bytes + 100 * 1024; // Add 100 KiB headroom to avoid reallocation.
+        Self {
+            buf_reader: BufReader::new(reader),
+            buffer: Vec::with_capacity(alloc_num_bytes),
+            alloc_num_bytes,
+            batch_num_bytes,
+        }
+    }
+
+    pub async fn next_batch(&mut self) -> io::Result<Option<Bytes>> {
+        loop {
+            let num_bytes = self.buf_reader.read_until(b'\n', &mut self.buffer).await?;
+
+            if self.buffer.len() > self.batch_num_bytes {
+                let mut new_buffer = Vec::with_capacity(self.alloc_num_bytes);
```

Comment:
This little buffer dance is a bit annoying. My initial thought on this was to use `memchr::memchr` directly instead of `read_until`

Reply:
Yeah, I agree. I was not keen on adding a new dependency, though.